src/streaming/array.writer.hh

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include "zarr.stream.hh"
+#include "thread.pool.hh"
+#include "s3.connection.hh"
+#include "blosc.compression.params.hh"
+#include "file.sink.hh"
+
+#include <condition_variable>
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
+namespace zarr {
+
+struct ArrayWriterConfig
+{
+    std::shared_ptr<ArrayDimensions> dimensions;
+    ZarrDataType dtype;
+    int level_of_detail;
+    std::optional<std::string> bucket_name;
+    std::string store_path;
+    std::optional<BloscCompressionParams> compression_params;
+};
+
+/// @brief Downsample the array writer configuration to a lower resolution.
+/// @param[in] config The original array writer configuration.
+/// @param[out] downsampled_config The downsampled array writer configuration.
+/// @return True if @p downsampled_config can be downsampled further.
+/// This is determined by the chunk size in @p config. This function will return
+/// false if and only if downsampling brings one or more dimensions lower than
+/// the chunk size along that dimension.
+[[nodiscard]] bool
+downsample(const ArrayWriterConfig& config,
```

Comment:
Why is `downsample` implemented as a free function rather than a public member? If it needs to be invoked outside the context of an `ArrayWriter` instance but is still related to it, it could be a static member. I don't think it's significant either way; I'm just trying to understand the reasoning behind this choice.

Reply:
I think we could probably (in later PRs) directly generate downsampled array writers without the intermediate step of generating configs, and get rid of this entirely.